### DISCUSSION TITLE
Transport layer security(TLS) for storage pods

### DISCUSSION THREAD

**Discussion by sunilkpmbl** [rook/rook#7220] [Q&A]
Dear team, 
I have been trying to find some specific information around securing pod to pod traffic for all the pods in the rook-ceph namespace using TLS and could not find any direct documentation around this. 
One approach we are trying at the moment is to use MTLS by adding Istio Sidecar to the pods but this is creating lots of other issues like the 'mon' pods are not reachable via service endpoints etc. (Posted an issue [here ](https://stackoverflow.com/questions/66182455/rook-ceph-pods-are-not-starting-properly-after-istio-sidecar-is-enabled?noredirect=1#comment117012525_66182455))

Is there a better way to do this?  Or is this already taken care internally?

**Comment by leseb** [comment]
Hi,

@sunilkpmbl it's not clear to me what is not working, especially because Rook does use Kubernetes Services for monitors by default.
I suspect it has something to do with Istio forwarding traffic. We need to dig a bit more.
You might consider enabling `secure` mode from Ceph's messenger V2. You can read more about it here: https://docs.ceph.com/en/latest/rados/configuration/msgr2/, also be careful with the performance impact associated with turning on encryption.
One more thing, @idryomov, is mgr2 secure mode supported with Ceph Kernel clients (rbd and cephfs)? If so, what are the kernel requirements (if any)? Thanks!

**Reply by idryomov** [reply]
msgr2 (including secure mode) is supported, but the most recent bits are needed: kernel 5.11 (just released) and some enablement work on the Ceph side, specifically in the rbd tool and filesystem mount helper.

**Reply by idryomov** [reply]
AFAIU mutual TLS provided by k8s means is supposed to be an interim solution, until everything trickles down into formal releases, including downstream.

**Reply by sunilkpmbl** [reply]
Thanks for the reply @leseb and @idryomov.

I have kind of narrowed down the issue to "rook-ceph-mon" pods. If we exclude sidecar injection on **rook-ceph-mon** and  **rook-ceph-osd-prepare** pods(this should be ok since its a one time scheduled job) things are working fine. 

In the Istio configuration I have added this to exclude **mon and prepare** pods from sidecar injection and it all worked after this. 

`  neverInjectSelector: 
    - matchExpressions:
      - {key: mon, operator: Exists}
    - matchExpressions:
      - {key: job-name, operator: Exists} `

The other thing I had to do was to make the MTLS mode "PERMISSIVE" from "STRICT".

When sidecar is enabled on rook-ceph-mon something strange was happening and it was not reachable through service endpoints.

I know this is a bit of a workaround. Looking forward for a better solution.

Happy to give more details if you want.

Full details of the issue was posted an issue [here](https://stackoverflow.com/questions/66182455/rook-ceph-pods-are-not-starting-properly-after-istio-sidecar-is-enabled?noredirect=1#comment117046099_66182455).

Also thanks for the hint on msgr2.
